diff --git a/content/browser/child_process_security_policy_impl.cc b/content/browser/child_process_security_policy_impl.cc
index bcd42e788019..a728123a1f31 100644
--- a/content/browser/child_process_security_policy_impl.cc
+++ b/content/browser/child_process_security_policy_impl.cc
@@ -1452,6 +1452,15 @@ bool ChildProcessSecurityPolicyImpl::CanAccessDataForOrigin(
 
     GURL actual_process_lock = security_state->origin_lock();
     if (!actual_process_lock.is_empty()) {
+      // ESP: check for and remove ephemeral context ID from the site hostname before doing checks
+      auto lock_host = actual_process_lock.host();
+      if (lock_host.compare(0, 6, "esp---") == 0) {
+        auto unlocked_host = lock_host.substr(41);
+        GURL::Replacements reps;
+        reps.SetHostStr(unlocked_host);
+        actual_process_lock = actual_process_lock.ReplaceComponents(reps);
+      }
+
       // Jail-style enforcement - a process with a lock can only access data
       // from origins that require exactly the same lock.
       if (actual_process_lock == expected_process_lock)
