diff --git a/content/browser/frame_host/frame_tree_node.h b/content/browser/frame_host/frame_tree_node.h
index 89ba68e856e9..2713a91bf848 100644
--- a/content/browser/frame_host/frame_tree_node.h
+++ b/content/browser/frame_host/frame_tree_node.h
@@ -112,6 +117,16 @@ class CONTENT_EXPORT FrameTreeNode {
     return devtools_frame_token_;
   }
 
+  // ESP: get stored navigation token
+  const base::Optional<base::UnguessableToken>& GetNavigationToken() const {
+    return devtools_navigation_token_;
+  }
+  // ESP: set stored navigation token
+  void SetNavigationToken(const base::UnguessableToken& navigation_token) {
+    devtools_navigation_token_ = base::make_optional(navigation_token);
+  }
+
+
   size_t child_count() const { return current_frame_host()->child_count(); }
 
   unsigned int depth() const { return depth_; }
@@ -528,6 +543,10 @@ class CONTENT_EXPORT FrameTreeNode {
   // if the frame is embedded in a parent frame.
   base::Optional<base::UnguessableToken> embedding_token_;
 
+  // ESP: stash of the last navigation token committed for this frame node
+  // (used only to track main-frame navigations for our top-level-document-lifespan keying)
+  base::Optional<base::UnguessableToken> devtools_navigation_token_;
+
   // Tracks the scrolling and margin properties for this frame.  These
   // properties affect the child renderer but are stored on its parent's
   // frame element.  When this frame's parent dynamically updates these
