diff --git a/content/browser/frame_host/frame_tree_node.h b/content/browser/frame_host/frame_tree_node.h
index 89ba68e856e9..327a1c6fd0ff 100644
--- a/content/browser/frame_host/frame_tree_node.h
+++ b/content/browser/frame_host/frame_tree_node.h
@@ -112,6 +112,13 @@ class CONTENT_EXPORT FrameTreeNode {
     return devtools_frame_token_;
   }
 
+  const base::Optional<base::UnguessableToken>& GetNavigationToken() const {
+    return devtools_navigation_token_;
+  }
+  void SetNavigationToken(const base::UnguessableToken& navigation_token) {
+    devtools_navigation_token_ = base::make_optional(navigation_token);
+  }
+
   size_t child_count() const { return current_frame_host()->child_count(); }
 
   unsigned int depth() const { return depth_; }
@@ -521,6 +528,10 @@ class CONTENT_EXPORT FrameTreeNode {
   // to look up the FrameTreeNode instance.
   base::UnguessableToken devtools_frame_token_;
 
+  // ESP: stash of the last navigation token committed for this frame node
+  // (used only to track main-frame navigations for our top-level-document-lifespan keying)
+  base::Optional<base::UnguessableToken> devtools_navigation_token_;
+
   // A token tracking the embedding relationship between a out-of-process
   // frames. This is only accessible by the renderer for the active
   // RenderFrameHost and the renderer for the RenderFrameProxyHost from
